clippy_lints/src/use_crate_prefix_for_self_imports.rs

```diff
@@ -0,0 +1,88 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_opt;
+use def_id::LOCAL_CRATE;
+use rustc_errors::Applicability;
+use rustc_hir::def::Res;
+use rustc_hir::{Item, ItemKind, def_id};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_session::declare_lint_pass;
+use rustc_span::{FileName, RealFileName};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// This lint checks for imports from the current crate that do not use the `crate::` prefix.
+    /// It suggests using `crate::` to make it clear that the item is from the same crate.
+    ///
+    /// ### Why is this bad?
+    /// When imports from the current crate lack the `crate::` prefix, it can make the code less readable
+    /// because it’s not immediately clear if the imported item is from the current crate or an external dependency.
+    /// Using `crate::` for self-imports provides a consistent style, making the origin of each import clear.
+    /// This helps reduce confusion and maintain a uniform codebase.
+    ///
+    /// ### Example
+    /// ```rust,ignore
+    /// // lib.rs
+    /// mod foo;
+    /// use foo::bar;
+    /// ```
+    ///
+    /// ```rust,ignore
+    /// // foo.rs
+    /// #[path = "./foo.rs"]
+    /// pub fn bar() {}
+    /// ```
+    ///
+    /// Use instead:
+    /// ```rust,ignore
+    /// // lib.rs
+    /// mod foo;
+    /// use crate::foo::bar;
+    /// ```
+    ///
+    /// ```rust,ignore
+    /// // foo.rs
+    /// #[path = "./foo.rs"]
+    /// pub fn bar() {}
+    /// ```
+    #[clippy::version = "1.84.0"]
+    pub USE_CRATE_PREFIX_FOR_SELF_IMPORTS,
+    style,
```

Comment:
While I understand that clearing up possible path confusion is a worthy cause, I'm afraid that introducing the lint as warn-by-default will lead to a lot of noise and broken CI jobs.

I'll discuss this on zulip.